src/Filecoin/Address.cpp

```diff
@@ -0,0 +1,196 @@
+// Copyright © 2017-2019 Trust.
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+#include "Address.h"
+
+#include "../Base32.h"
+
+#include <TrezorCrypto/blake2b.h>
+
+using namespace TW::Filecoin;
+
+static const char BASE32_ALPHABET_FILECOIN[] = "abcdefghijklmnopqrstuvwxyz234567";
+static constexpr size_t checksum_size = 4;
+
+// Calculate the checksum of a byte slice.
+static std::array<uint8_t, checksum_size> calc_checksum(const std::vector<uint8_t>& bytes) {
+    std::array<uint8_t, checksum_size> sum;
+    blake2b_state hash;
+    blake2b_Init(&hash, checksum_size);
+    blake2b_Update(&hash, bytes.data(), bytes.size());
+    blake2b_Final(&hash, sum.begin(), sum.size());
+    return sum;
+}
+
+bool Address::isValid(const std::vector<uint8_t>& data) {
+    if (data.size() < 2)
+        return false;
+    auto type = data[0];
+    if (type == address_type_id) {
+        // Verify varuint encoding
+        if (data.size() > 11)
+            return false;
+        if (data.size() == 11 && data[10] > 0x01)
+            return false;
+        int i;
+        for (i = 1; i < data.size(); i++)
+            if ((data[i] & 0x80) == 0)
+                break;
+        return i == data.size() - 1;
+    } else {
+        return data.size() == 1 + Address::payload_size(type);
+    }
+}
+
+bool Address::isValid(const std::string& string) {
+    if (string.length() < 3)
+        return false;
+    // Only test net addresses supported.
+    if (string[0] != 't')
+        return false;
+    // Get address type.
+    uint8_t type;
+    switch (string[1]) {
+    case '0':
+        type = address_type_id;
+        break;
+    case '1':
+        type = address_type_secp256k1;
+        break;
+    case '2':
+        type = address_type_actor;
+        break;
+    case '3':
+        type = address_type_bls;
+        break;
+    default:
+        return false;
+    }
+    uint8_t payload_size = Address::payload_size(type);
+
+    // ID addresses are special, they are just numbers.
+    if (type == address_type_id) {
+        if (string.length() > 22)
+            return false;
+        try {
+            std::stoull(string.substr(2));
+        } catch (...) {
+            return false;
+        }
+        return true;
+    }
+
+    // Check if valid Base32.
+    Data decoded;
+    if (!Base32::decode(string.substr(2), decoded, BASE32_ALPHABET_FILECOIN))
+        return false;
```

Comment:
```suggestion
        { return false; }
```
always wrap with `{}`